### DISCUSSION TITLE
Pub - Sub model on beckn

### DISCUSSION THREAD

**Discussion by ravi-prakash-v** [beckn/protocol-specifications#189] [Ideas]
Can we include a pub-sub model in the beckn protocol specification? i.e Event-driven over the usual message-driven approach. 

Each network participant can implement a BAP message queue with topics for `search`, `select`, `init` etc; and BPPs can subscribe to these topics. similarly BPPs can publish their responses on a BPP message queue with topics for `on_search`, `on_select`, `on_init` etc and BAPs can subscribe to them. 

Supported technologies could be MQTT, AMQP, Cloud Events.

It would be nice if someone can publish an architecture and reference implementation for such a network.

**Comment by gamemaker1** [comment]
Hi @core-wg-admin,

I am interested in contributing to this. I could do the following:

1. Create an Async API specification for the Beckn Protocol.
2. Create (or extend the current one) a reference implementation for a pub sub network following the event-driven Beckn Protocol.

I'll keep posting the updates to this thread so everyone can contribute/provide comments.

Thanks,
Vedant

**Reply by Vedsaga** [reply]
I would love to volunteer in this😇 if I get opportunity...

**Comment by Vedsaga** [comment]
Hii I am really excited to contribute to this,

>  I will be able to share architecture digram specially based around AMQP &  MQTT

However,  I have qurey which is to create pub-sub model based on MQTT it does involve broker and I am gussing is that there can two kind of MQTT model one which is `Open MQTT` which can be accessed by any BPPs `[this will come handy when it comes to implementing or building GIN(Global Inventory Network)]` another would be `closed MQTT` which can created by networks by collectively coming together.  So, my question is in case of close MQTT network participant will have to pay for its maintenance however.



> `So in open MQTT who will be responsible broker (it can one ore more)  to maintain the open MQTT?`

**Reply by venkatramanm** [reply]
I imagine it will closed to begin with and be opened when ripe enough. Open broker is most inclusive and useful in the long run.. yes who will maintain, fees etc are valid questions. May be, it could become fees added to your broadband bill at some point. Any thing is possible. Love beckn for opening up these transformational possibilities. 

Lets brainstorm more.

**Reply by Vedsaga** [reply]
yee, surely :D

**Comment by venkatramanm** [comment]
I can volunteer in this project to crystalize spec, network policy and   building the reference implementation. 

It is a great initiative.

**Comment by gamemaker1** [comment]
Hi @vedsaga @venkatramanm,

I would love to collaborate on you with this :D

I have created a repository (https://github.com/gamemaker1/beckn-async-api-spec) where we can collaborate on the Async API specification for the event-driven version of the Beckn Protocol and invited you as collaborators.

Once we get a draft of the Async API ready, we could proceed to create a reference implementation. Does that sound good?

Looking forward to collaborating with you :D

**Reply by Vedsaga** [reply]
Sounds Good to me 😇

**Reply by sujitrs** [reply]
Possible Pub-Sub with current Beckn Protocol. This was done using Solace in Mind. But its costly. Hence i can collaborate to design it with FOSS tech.
<img width="834" alt="(Producer" src="https://user-images.githubusercontent.com/23429265/149261899-bdda0d82-3c61-4390-8ec6-8d73420425a7.png">

Here Producer (A) can be BAP and Consumer (F) be BPP for `search`
And for `on_search` reverse.

**Reply by Vedsaga** [reply]
pub-sub model is only suppose to act as broker that hold on the messages and pass on to the right BPPs and rest of the other job have to done by BPPs themselves?

**Reply by gamemaker1** [reply]
Hi @sujitrs,

Thanks for taking the time to create and share this architecture diagram, it is quite helpful!

A few questions that I had:

- What if we removed the Beckn Application's call to the API gateway (B) and instead allowed it to directly publish a message to the queue which the Beckn Providers will be subscribed to? The same could go for F - the BA is already subscribed to the topic and can listen continuously for responses.
- Could D be treated as the Beckn Provider - it will receive the request from the `action` channel and then publish the response on the `on_action` channel (where `action` can be `search`, `init`, `confirm`, etc.). I believe @Vedsaga had the same question (https://github.com/beckn/protocol-specifications/discussions/189#discussioncomment-1964057)
- Instead of creating topics on the fly, what if we had only a predefined set of topics, i.e., `action` and `on_action` (where `action` can be `search`, `init`, `confirm`, etc.). The BAs will publish requests on the `action` channel, and the BPs will respond to them on the `on_action` channel. The BP can use the same `context.message_id` while responding so the BA knows that the message is in response to its request.

Please let me know your thoughts on the above.

I have also invited you as a collaborator to the https://github.com/gamemaker1/beckn-async-api-spec repository where we are building the spec and implementation. Looking forward to collaborating with you on this :D

**Reply by Vedsaga** [reply]
>     * Instead of creating topics on the fly, what if we had only a predefined set of topics, i.e., `action` and `on_action` (where `action` can be `search`, `init`, `confirm`, etc.).

I prefer the idea of having pre define topic because it will enable subscriber to reach out to the exact broker for it's messages instead of query having to going to alot of searches.

**Comment by Vedsaga** [comment]
# idea
 Two-more beckn networks share & exchange data on occurrence of an event using pub-sub model

# Key Elements in a MQ pub-sub model
1. publisher - `one who publish the message`
2. principal broker - `one who distribute the publisher messages between the event broker `
3. event broker - `one who is responsible to manage topics of an event`
4. topic - `used to classify the message`
5. subscriber - `one who consume messages`
6. message - `contain the event description`

# prerequisite(proposal)
1. subscriber identity would be it’s protocol ID which was assign during creation of network which is stored in registery so, that while making query it can bind that query with it's identity so that correct messages can be delivered.
2.` messages should have status (delivered/failed/partially-distributed)` so that, in case  message was **_o2m(one to many network)_** & on first attempt message didn't get deliver to all the subscriber then it can be marked as `partially-distributed` . 
3.` message should have time-out period` so, that during that period re-attempt can be performed.
4. `message should have defined re-attempt` that a distributor will try, if all the attempt is consumed and subscriber don't send with acknowledgement then distributor will mark the `message status to failed` if in case message type was _**o2o(one to one network)**_,  in case _**o2m(one to many network)**_ then status will be marked as `partially-distributed` ; and then acknowledgement will be to the publisher 
5. `message should have defined type (all/group/binded)`. 
- message with type all will be shared with/accessible by all subscribers of that topic.
- message with type  group will be shared with/accessible by only those subscriber who are member of that group.
- message with type  binded will be only accessible by subscriber that matching id mentioned in that message.
6. message should only be deleted if message status changes to `delivered` or `message timeout period is over`.

# diagram 
`Note: below diagram just only represent distribution of binded messaged.`

I have attached the PDF of diagram in case image not clear.
![direct-binded-messages-diagarm-black](https://user-images.githubusercontent.com/59083528/149376396-94a881f0-a3cd-487e-b555-9f4ddac268b5.png)
[direct-binded-messages-diagarm-white.pdf](https://github.com/beckn/protocol-specifications/files/7864597/direct-binded-messages-diagarm-white.pdf)
[direct-binded-messages-diagarm-black.pdf](https://github.com/beckn/protocol-specifications/files/7864600/direct-binded-messages-diagarm-black.pdf)

**Reply by gamemaker1** [reply]
This is a great implementation architecture, thanks for sharing!

A few questions that I had:

- "A message with type `group` will be shared with/accessible by only those subscriber who are member of that group": How are groups different from topics? 
- "A message with type `binded` will be only accessible by subscriber that matching id mentioned in that message.": How would you send messages to a particular Beckn Application?
- How would an acknowledgment be sent to the entity in a pub sub implementation?
- How would the status of the message be changed and who would change it?

I think the last two questions would be partly answered if we chose how we want to implement pub-sub, i.e., using Webhooks, Websockets, MQTT, or something else?

Please let me know your thoughts on this.

**Reply by Vedsaga** [reply]
>   How are groups different from topics?

I would to like answer this but, I think I have to make diagram to explain the reason for having groups and what is the difference between the topics & groups. Please allow me some time I will  answer about that in depth. Because to understand it  I think we have deep dive into the type relationship between entities. Reason for having groups is because this way we can have form `connection between the networks `. Just like how internet work as there are different of network topology which enable data sharing between different computers  same way if can have groups then we can allow networks to constantly exchange data with each other. Which will enable grater possibility instead of just mare  having discovery base system. I have open discussion on this idea  #181. I am still working on that idea. 

>  How would you send messages to a particular Beckn Application?

In PR which you propose in meta data sender have to attach the binding key of that specific receiver. That binding key could be the same key unique key which is currently provided by registery. Now, whenever reader want to check with pub-sub either there is any messages or not for them they have to attach their private key so, that `verifier` can make sure that it is valid query so, if it is valid then distributor will check the DB and if there is any message for that entity then it will forward those messages to that entity. Now, here we have make sure that whenever message is forwarded it can't interpreted in between. Similar to how SMTP. When a user of Gmail send email to outlook user it only receive by outlook user and not by anyone in between. 

>     * How would an acknowledgment be sent to the entity in a pub sub implementation?

So, each message which is send by sender they will attach their signature that are the one who published it. So, there is something call `save` which will read the `meta data` and able to _extract the signature  of the sender_ and then it will pass the request to `database maintainer `to save the message and as soon as `sender` receive the confirmation it will send the ACK to that `signature  address`. So, what I am trying say is that `signature of sender` will refer as sender address.

>     * How would the status of the message be changed and who would change it?

These have to be done by the topic-broker who is responsible to manage topic's events. I want clarify one thing that is Each topic such as Shipment can have multiple events _(here event means any message which an entity wants to share with another)_.  For e.g. when there is order for long distance delivery  like From Delhi To Chennai a  logistic BPP truck can be stop at multiple points. And because it is long distance delivery there could be chance of having multiple logistic BPP involve in this. One logistic BPP for inter city transportation  like `DELIVERY  - India's biggest fulfillment startup` & another BPP such as `DUNZO - operate at city level ` for city level delivery. Now, first  logistic BPP might want to share data with Buyer BAP, Seller BPP & another  Logistic BPP at the same time. Now, for a moment imagine that  DELIVERY  company truck stop at warehouse which is in Indore. Now, this have to share with all the stakeholder which is involved in the order right? and so  DELIVERY  company sent the message to the `pub-sub` which is suppose to distribute the copy of the message with all the stakeholder. Now, what will happen is `pub-sub` will create copy of the message and then will try to send the message to the all the stakeholder. But due to some reason Seller BPP server is down or for some reason `pub-sub` is not getting ACK from that them. That means `pub-sub` can't discard the message from the db either `TTL` of message is over or `till it does not receive ACK from all the stakeholder` and so, due to that reason we need to keep track of `message status ` so that when messages are discarded from the DB sender can know exactly what happen with it's messages. Either it is even read by stakeholder or not.


> I think the last two questions would be partly answered if we chose how we want to implement pub-sub, i.e., using Webhooks, Websockets, MQTT, or something else?

The above model is similar is to RabbitMQ but, with few enhancement because we are tying to implement it in open environment where multiple independent sever will communicate, as currently RabbitMQ is being used by companies to handle MQ but in close environment.

**Comment by gamemaker1** [comment]
Hi @Vedsaga @venkatramanm @sujitrs,

I have opened a draft pull request (https://github.com/gamemaker1/beckn-async-api-spec/pull/1) on the Async API specification repo to add the `search` and `on_search` channels to the spec. Please do review it and provide your feedback.

Thanks,
Vedant

**Comment by AnshulMalik** [comment]
Hi @core-wg-admin 

If I have understood correctly, we are trying to bring one central dependency of a message broker in this decentralised network.

And I can see few concern if we go this route:
1. This is adding a single point of failure.
2. How are we going to handle the scale?
    2.1 When someone subscribes to a topic, will it get ALL the events? when in reality, it only needs just one.
3. What advantages are we trying to gain with this approach?

**Reply by Vedsaga** [reply]
>If I have understood correctly, we are trying to bring one central dependency of a message broker in this decentralized network.

I think we will going to have multiple message broker and it can be not just going to be open but, it can be close too between multiple networks .

> What advantages are we trying to gain with this approach?

For an e.g. - there is group of BPPs who operate in different region but, they are depend on each for sharing of updates on occurrence of an event like they have received a order from seller BPP. Now, currently first  BPP can inform it's partner by  going through partners-list and sending copy to each of them. But, imagine a case where a BPP want to share the information to all the BPPs about an event in that case a BPPs have to know about all the BPPs who operate in that domain. Now, goal is to decrease the infrastructure work for the participants that they have to build & maintain anyway but this will enable them to share same infra structure. Not just that it will enable another use case so far example currently the buyer Bap or seller bpp have  to build the component to let user see the delivery updates right and this how it is being done currently, as all the deliver companies have there own website user can see the delivery updates about it's order. With this there could be a single  global component with which all logistics BPPs can be connected and user can come here give it's order id and can see the real time updates.
>     1. This is adding a single point of failure.
This is true, but chances of failure can be decrease by not building one broker for all the events but, each broker for events to manage topics.
>     2. How are we going to handle the scale?
This is tricky question because agree we are talking about global scale , may be by having multiple instance for each broker and I surely believe we can tackle this challenge by working together on this.
>        2.1 When someone subscribes to a topic, will it get ALL the events? when in reality, it only needs just one.
Nope, a subscribes  will only get receive all the events. Firstly, subscriber will have subscribe to the events if they have not then all the message that this broker will receive will discard it and not save in the queue. And, not just that only the message with type `message with type all` will be share rest other two message type `group, binded` will only be accessible by those members who suppose to receive it.

I have propose about this idea in the above comment.

**Reply by rahul101001000** [reply]
Agree with the points @Vedsaga 

> How do we handle scale?

If at all it needed to be done, it's fairly common to scale message brokers. Do take a look at how it is done with [Netflix device management platform](https://netflixtechblog.com/towards-a-reliable-device-management-platform-4f86230ca623) and in the AdTech world at [AdsDax](https://hedera.com/blog/how-adsdax-built-a-scalable-decentralized-message-queue-with-hedera-hashgraph). 

With Beckn as @Vedsaga mentioned, just like there will be multiple regional instances of Beckn gateways, so will there be multiple instances for message brokers.

And yes @AnshulMalik  all great questions - that's exactly why we can together try and solve for them. @Vedsaga @gamemaker1 @sujitrs @venkatramanm have good ideas in there to make this work. Once we have a prototype, we can try out a few use cases / test cases and evaluate it. If all good, we could add it to the core spec.

**Reply by AnshulMalik** [reply]
> Beckn is a protocol, not a platform. It adopts a decentralized architecture that obviates the need for creating a centralised platform in order to integrate services from multiple providers simultaneously ensuring privacy and security by design by enabling secure, encrypted iteractions.

This is something I found in the readme. I feel like we are moving away from our goal here, we were trying to make a decentralised system. And additionally, this change would restrict people on ONE pub-sub implementation.

**Reply by rahul101001000** [reply]
Let me phrase the question in different ways - 
1. How would we implement Beckn with AsyncAPI without a message broker?
2. How would we enable this use case in Beckn without pubsub - "Every order placed, with user consent, is viewed by a pool of OCEN providers that see the order, determine credit worthiness, and publish an event viewable by the BAP (and the BPP where purchase is being made) that indicates that they (OCEN provider) are willing to offer an EMI option for that user for that purchase"
3. Why do you say message broker is centralized. Message brokers will be as decentralized as Beckn gateway.

**Reply by AnshulMalik** [reply]
Hey @rahul101001000 


> How would we implement Beckn with AsyncAPI without a message broker?

I am assuming that the idea of migrating to AsyncAPI spec came because we don't want to be limited to HTTP, but if there is no other good way of implementing AsyncAPI then it's one and the same thing again no? Earlier it was HTTP that was fixed, now it's the broker protocol that's fixed, we have just changed the fixed constraints to be some other constraints.


> How would we enable this use case in Beckn without pubsub?

I am assuming BPP publishes the event on order creation to relevant stakeholders. I can also see security and privacy concerns here if a message is visible to some entity where it shouldn't be and how are we going to achieve this control in broker based architecture?
How do we currently handle this case?

> Why do you say message broker is centralized. Message brokers will be as decentralized as Beckn gateway.

Sorry, I am not able to visualize this, Could you help me visualize how would a search and order operation would work with this decentralized broker?

**Reply by Vedsaga** [reply]
> I am assuming BPP publishes the event on order creation to relevant stakeholders. 

Not just BPP but, BAP can also be publisher here. Idea is that both BAP & BPP can be publisher as well subscriber at the same time.

> I can also see security and privacy concerns here if a message is visible to some entity where it shouldn't be and how are we going to achieve this control in broker based architecture? How do we currently handle this case?

It is a valid concern, however to tackle that we can take inspirational approach which RabbitMQ have it. That is  it first classify the message into `access type `  --
 1. Fanout(can be access by every subscriber)
 2. Binded(can access by only that subscriber who have been refer into that message by publisher)
 3.  Partial Bindied (message that have publish for a group of networks that means we will have system which will keep trace of groups and group memeber will be assigned with a unique key which they can use to access the group messages)
 
> Why do you say message broker is centralized. Message brokers will be as decentralized as Beckn gateway.
> Sorry, I am not able to visualize this, Could you help me visualize how would a search and order operation would work with this decentralized broker?

It will not be completely decentralized because there will be multiple entity who will be responsible to maintain each event-broker however because a event-broker not working  1 network but, for all the network that is why it is not completely centralized. it is a mix of decentralized & centralized system.
